design-documents/Sale_Payment_Operation.md

```diff
@@ -0,0 +1,120 @@
+![](sale_payment_operation_img/order_payment_new.png)
+
+`OrderPayment` will call `SaleOpertion` if `sale` operation is available for payment method. Otherway, `CaptureOperation` will be called for backward compatibility with existing payment integrations.
+
+**Magento\Sales\Model\Order\Payment**
+```php
+<?php
+protected function processAction($action, Order $order)
+{
+    $totalDue = $order->getTotalDue();
+    $baseTotalDue = $order->getBaseTotalDue();
+    
+    switch ($action) {
+        ............................................................................
+        case \Magento\Payment\Model\Method\AbstractMethod::ACTION_AUTHORIZE_CAPTURE:
+            $this->setAmountAuthorized($totalDue);
+            $this->setBaseAmountAuthorized($baseTotalDue);
+            if ($this->canSale()) {
+                $this->saleOperation->sale($this);
+            } else {
+                $this->capture(null);
+            }
+            break;
+        default:
+            break;
+    }
+}
+ 
+/**
+ * Check sale operation availability for payment method.
+ *
+ * @return bool
+ */
+private function canSale(): bool
+{
+    $method = $this->getMethodInstance();
+    return $method instanceof SaleOperationInterface && $method->canSale();
+}
+```
+
+## Payment provider integration configuration
+Payment integration should be based on [Magento payment provider gateway](https://devdocs.magento.com/guides/v2.2/payments-integrations/bk-payments-integrations.html).
+
+Integrations based on deprecated `Magento\Payment\Model\Method\AbstractMethod` this solution doesn't support.
+
+In `config.xml` `can_sale`  option should be defined:
+
+```xml
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Store:etc/config.xsd">
+    <default>
+        <payment>
+            <braintree>
+                ......................
+                <can_sale>1</can_sale>
+                ......................
+            </braintree>
+        </payment>
+    </default>
+</config>
+```
+In `di.xml` `sale` command should be defined in `CommandPool`:
+```xml
+<!-- Commands infrastructure -->
+<virtualType name="BraintreeCommandPool" type="Magento\Payment\Gateway\Command\CommandPool">
+    <arguments>
+        <argument name="commands" xsi:type="array">
+            ...............................................................
+            <item name="sale" xsi:type="string">BraintreeSaleCommand</item>
+            ...............................................................
+        </argument>
+    </arguments>
+</virtualType>
+```
+
+## Backward compatibility
+ - New @api interface `SaleOperationInterface` should be introduced - **MINOR** change according to SemVer
+ - Two public methods should be added in `Magento\Payment\Model\Method\Adapter` @api class - *sale()* and *canSale()*
```

Comment:
These methods will have implementation as far as I understand, if so - this will be minor change, otherwise it will be a major change.

Reply:
The base implementation of `sale()` and `canSale()` will be in `Magento\Payment\Model\Method\Adapter` as it is the main extension point for payment integrations.

Reply:
Yes, they will